### PR TITLE
WIP: Align stack size with long size and add generators

### PR DESCRIPTION
**Pull request by vilu**
Just opening this up for some feedback, will continue to clean up and push to this.

### CODE REVIEW

File: core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala [hunk 1]
```diff
@@ -27,30 +27,32 @@ private[zio] final class StackBool private () {
   private[this] var _size = 0L
 
   final def getOrElse(index: Int, b: Boolean): Boolean = {
-    var j   = index.toLong
+    val i0  = _size & 63L
+    val i   = ((64L - i0) + index) & 63L
+    var ib  = ((64L - i0) + index) / 64L
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala]:
```suggestion
    var ib  = ((64L - i0) + index) >> 6
```

Comment (vilu) [reply core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala]: Thanks for the input, just out of curiosity. Why do you think it's better?

**Comment by vilu** [comment]
I've added generators (I'll clean up later) however, seems like the large push/pop test broke. Will have to check that implementation too.

File: core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala (hunk 1 above)
Comment (regiskuckaertz) [reply core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala]: I think bitwise shift is much faster than division. You can as well avoid computing `(64L - i0) + index` twice 😃

**Comment by vilu** [comment]
Ok, didn't know that. Thanks. I'll clean up the code when I got everything
working and I'll make sure things are only calculated when necessary.

Den lör 25 maj 2019 kl 14:41 skrev Regis Kuckaertz <notifications@github.com

File: core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala [hunk 2]
```diff
@@ -27,30 +27,32 @@ private[zio] final class StackBool private () {
...
     var cur = head
-    while (j >= 256L && (cur.next ne null)) {
-      j -= 256L
+
+    while (ie > 0) {
+      ie -= 1
       cur = cur.next
     }
-    assert(j < 256L && j >= 0)
+    assert(i < 64L && i >= 0)
```
Comment (tkroman) [inline core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala]: isn't i always going to be < 64?

**Comment by vilu** [comment]
I kept this part because it was there before. In theory it should never blow up.

**Comment by tkroman** [comment]
yes, i mean there is no need for assert on this because `i` is immutable, there was an assert before on `j` because it changes in the loop, so maybe you want to change to assert on the value of `ie` (the variable one)?

**Comment by vilu** [comment]
I ended up removing the assertion. Thanks for pointing it out.

File: core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala [hunk 3]
```diff
@@ -27,30 +27,31 @@ private[zio] final class StackBool private () {
   private[this] var _size = 0L
 
   final def getOrElse(index: Int, b: Boolean): Boolean = {
-    var j   = index.toLong
+    val i0  = _size & 63L
+    val i   = ((64L - i0) + index) & 63L
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala]:
```suggestion
    val base = (64L - i0) + index
    val i    = base & 63L
    var ie   = base >> 6
```

File: core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala [hunk 4]
```diff
@@ -27,30 +27,32 @@ private[zio] final class StackBool private () {
...
+    var ie  = ((64L - i0) + index) >> 6
     var cur = head
-    while (j >= 256L && (cur.next ne null)) {
-      j -= 256L
+
+    while (ie > 0) {
+      ie -= 1
       cur = cur.next
```
Comment (LGLO) [inline core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala]:
```
val stack = StackBool()
stack.getOrElse(300, false)
```
NullPointerException

File: core/shared/src/test/scala/scalaz/zio/internal/StackBoolSpec.scala [hunk 5]
```diff
@@ -1,78 +1,61 @@
...
+    Prop.forAll(generator) { list: List[Boolean] =>
+      val stack = StackBool()
 
-    list.foreach(stack.push(_))
+      list.foreach(stack.push(_))
 
-    list.reverse.foldLeft(true must_=== true) {
-      case (result, flag) =>
-        result and (stack.popOrElse(!flag) must_=== flag)
+      list.reverse.foldLeft(true must_=== true) {
```
Comment (LGLO) [inline core/shared/src/test/scala/scalaz/zio/internal/StackBoolSpec.scala]: `true must_=== true` - I was looking for something like `MatchSucceed` but couldn't find.

File: core/shared/src/test/scala/scalaz/zio/internal/StackBoolSpec.scala [hunk 6]
```diff
@@ -1,78 +1,61 @@
...
+        Peek/pop identity             $e3
     """
 
-  def e0 = {
-    val list = List.fill(100)(true)
+  import Arbitrary._
 
-    StackBool(list: _*).size must_== list.length
-  }
+  private val generator: Gen[List[Boolean]] = boolListGen(0, 400)
```
Comment (LGLO) [inline core/shared/src/test/scala/scalaz/zio/internal/StackBoolSpec.scala]: I might be wrong, but maybe 200 is big enough - there are 3 entries at this point.

File: core/shared/src/test/scala/scalaz/zio/internal/StackBoolSpec.scala (hunk 5 above)
Comment (vilu) [reply core/shared/src/test/scala/scalaz/zio/internal/StackBoolSpec.scala]: Not sure if this is something I should act on based on your comment? I followed the already existing pattern here. I haven't used specs2 before so I'm unaware of best practices in this case.

File: core/shared/src/test/scala/scalaz/zio/internal/StackBoolSpec.scala (hunk 6 above)
Comment (vilu) [reply core/shared/src/test/scala/scalaz/zio/internal/StackBoolSpec.scala]: I've adjusted the max size according to your recommendations.

File: core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala (hunk 3 above)
Comment (vilu) [reply core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala]: Thanks, I was struggling with the naming here :)

File: core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala (hunk 4 above)
Comment (vilu) [reply core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala]: Thanks for spotting this. I've added a null check here.

**Comment by vilu** [comment]
@LGLO, @regiskuckaertz thanks for your feedback. I've tried to address your comments.

**Comment by vilu** [comment]
> For "laws" I like ScalaCheck, but we still can have simple checks for some scenarios, like getting element with too high index.

@LGLO Can you think of any other case besides too high index (I've added one check there)?

File: core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala [hunk 7]
```diff
@@ -70,17 +72,17 @@ private[zio] final class StackBool private () {
...
-        if (index == 0L && head.next != null) head.next else head
+        if (index == 63L && head.next != null) head.next else head
 
       ((1L << index) & entry.bits) != 0L
     }
 
   final def popDrop[A](a: A): A = { popOrElse(false); a }
 
   final def toList: List[Boolean] =
-    (0 until _size.toInt).map(getOrElse(_, false)).toList.reverse
```
Comment (LGLO) [inline core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala]: IDK if there are any laws that bind stack and list, but my intuition is that `toList` should yield same result (besides mutation) like `pop`-ing all elements. However non-mutating `toList` is just inefficient and I consider this as "debug" method only.

File: core/shared/src/test/scala/scalaz/zio/internal/StackBoolSpec.scala (hunk 5 above)
Comment (LGLO) [reply core/shared/src/test/scala/scalaz/zio/internal/StackBoolSpec.scala]: No action required from my side :+1:

**Comment by LGLO** [comment]
@vilu From my side it looks OK! I don't merge right now because I'd like to take another look but I don't have time today. I'll take a look tomorrow.

File: core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala (hunk 7 above)
Comment (vilu) [reply core/shared/src/main/scala/scalaz/zio/internal/StackBool.scala]: Yes, I was thinking about that too. I had a quick look around and I was seeing different behaviours which made me think that it should just continue the way it behaved before.



`
import java.util.Stack;
import java.util.ArrayList;

public class Main {
    public static void main(String args[]) {
        Stack<Boolean> stack = new Stack<Boolean>();
        stack.push(true);
        stack.push(true);
        stack.push(false);
        stack.push(true);
        stack.push(false);
        ArrayList<Boolean> list = new ArrayList(stack);
        for (int i = 0; i < list.size(); i++) {
          System.out.println(list.get(i));
        }
    }
}

// true
// true
// false
// true
// false
---------------------------------
scala> res27
res29: s.type = Stack(false, true, false, true, true)

scala> res27.toList
res30: List[Boolean] = List(false, true, false, true, true)
---------------------------------
val s = StackBool()

s.push(true)
s.push(true)
s.push(false)
s.push(true)
s.push(false)


println(s.toList)
//List(false, true, false, true, true)
`

**Comment by jdegoes** [comment]
@vilu Thank you!

**Pull request merged**